### DISCUSSION TITLE
Apostrophe 3.30.0: Vue3 support, improved UI, and code linting

### DISCUSSION THREAD

**Discussion by BoDonkey** [apostrophecms/apostrophe#3913] [Release Notes]
## Bonjour Apostrophe Lovers!
This release cycle focused on some core improvements and a lot of tech design for our next features.

There were three notable core improvements - see the changelogs below for details.

<dl>
<dd>First, while Apostrophe’s admin UI remains in Vue 2, you can now use Vue 3 as part of a custom, project-specific webpack build for the public parts of your UI, if you choose. While the core admin UI is still written in Vue 2, both versions can co-exist in your project without the conflicts that were formerly reported during the webpack build process.</dd>

<dd>Second, the `APOS_LOG_ALL_ROUTES` environment variable has been introduced to allow debugging of middleware functions and routes.</dd>

<dd>Third, the `attachment` module has a new option and method for adding or extending file uploads in answer to multiple requests from the community.</dd>
</dl>

---

### New Advanced Permission UI

<img width="1437" alt="The new UI od the add group modal of the advanced permission module" src="https://user-images.githubusercontent.com/13168224/195434263-e4f2b420-ab65-4fec-a587-4c2e034c5af4.png">
The `advanced-permission` module got a UI update, with a more intuitive and easy-to-read grid replacing the former piece-type groupings.

---

Once again, for those in the back that didn't hear it last release, it is still Hacktoberfest. We welcome pull requests! This release cycle also features updates to our team [stylelint configuration](https://www.npmjs.com/package/stylelint-config-apostrophe). We welcome anybody making open-source contributions to our repos to use this configuration.

---

The ❤️ for editors will be back again in the next cycle with on-page placeholder content to help Editors visualize what pages will look like and keyboard shortcuts to help them get the job done quicker!

Until the next release, remember, it's not too early to start planning for 19 Jan 2038!
<p>&nbsp;</p>

---

### Adds

* New `APOS_LOG_ALL_ROUTES` environment variable. If set, Apostrophe logs information about all middleware functions and routes that are executed on behalf of a particular URL.
* Adds the `addFileGroups` option to the `attachment` module. Additionally, it exposes a new method, `addFileGroup(group)`. These allow easier addition of new file groups or extension of the existing groups.

### Fixes

* Vue 3 may now be used in a separate webpack build at project level without causing problems for the admin UI Vue 2 build.
* Fixes `cache` module `clear-cache` CLI task message
* Fixes help message for `express` module `list-routes` CLI task
<p>&nbsp;</p>

## Apostrophe 3.x modules

[`@apostrophecms/stylelint-config-apostrophe`](https://www.npmjs.com/package/stylelint-config-apostrophe) 2.0.0

This module provides a stylelint configuration for the Apostrophe team and contributors.

### Adds

- Update dependencies and rules for stylelint v14+, remove `peerDependencies` package property.
<p>&nbsp;</p>

## Enterprise modules

`@apostrophecms-pro/advanced-permission` 1.2.0

> Note: this module allows admins much greater power in creating and assigning permissions to custom groups, but this requires license upgrade. [Contact us](https://apostrophecms.com/contact-us) learn more.

### Changes

* Add a new permission grid that allows to edit and view the effective permissions in a single table. All content types
    are listed in the table by default, the `Add Role` button has been removed in consequence.
* Remove `All` type permission. This improves security when new types are added and avoids confusion since 'All' never really included everything. A one-time migration grants groups that used to have 'All' explicit permissions for the same types.
<p>&nbsp;</p>

`@apostrophecms-pro/document-versions` 1.0.2

> Note: this module automatically creates versions for your published documents (pages and pieces) and allows manual restore to any previously saved document version. This module requires a license upgrade. [Contact us](https://apostrophecms.com/contact-us) learn more.

### Changes

- Removed confusing boilerplate from README.

**Comment by iojanis** [comment]
How to use Vue in the public part of the app?

**Reply by boutell** [reply]
Hi iojanis,

What we did in this release is make sure Apostrophe stays out of your way by not causing an npm-related error if you make vue 3 a dependency of your project and create your own webpack build for it, or perhaps use our webpack extension support to enable it for .vue files. We don't currently have an example to share, but wanted to make sure no one is blocked from using this combination of features.
